#158175316 Notify the user when a job is running too often.

RrNn opened the pull request:
### What does this PR do?
- Ensure that the user is notified when a job is running too often
### How should this be manually tested?
- Clone the branch, `git clone https://github.com/andela/hc-wits-kla/tree/ft-user_notification-158175316`
- In your terminal, `cd hc-wits-kla`
- In the project directory, activate the environment, `source .env`
- Install requirements `pip install requirements.txt`
- Make migrations `python manage.py makemigrations`
- Run migrations `python manage.py migrate`
- Serve the application `python manage.py runserver`
- In the browser, visit, `127.0.0.1:8000/checks/`
- In your terminal, listen for alerts `python manage.py sendalerts`
- In your browser create a check, set **Period** to 2 minutes and **Grace Time** to 1 minute  and copy-paste it in another tab, and then again in another tab, hitting enter to load the page in each case. Do this in less than a minute.
- Go to the console and wait, the out put should be` -- PINGED QUITE OFTEN --`
### Relevant Pivotal Tracker story.
[#158175316](https://www.pivotaltracker.com/story/show/158175316)

Yiga-fred reviewed (commented):
Good work but try following the team's git workflow  by writing commit messages following the recommended pattern, have a look here #### https://github.com/andela/bestpractices/wiki/Git-naming-conventions-and-best-practices

RrNn commented:
Okay, @solnsubuga and @Yiga-fred . I'm gonna edit the message shortly.

wasswa-derick reviewed (approved):
All reviews seem okay.
For commit messages, in the best practices guide, somewhere it gives guidance on the header, body, and other footnotes. It shall help for long commit messages.

solnsubuga reviewed (dismissed):
@RrNn you accomplished the task of determining when a check is pinged too often well. 
<img width="602" alt="screen shot 2018-07-07 at 00 51 26" src="https://user-images.githubusercontent.com/23418080/42401843-36fe4328-8180-11e8-8577-b34e419c788f.png">
Though I configured email settings locally and realised that  you do not notify the user with an alert showing which check is being pinged too often. I suggest after you identify the check, trigger an alert to be sent to the user.

john555 commented on hc/api/management/commands/sendalerts.py:
```diff
@@ -68,3 +69,39 @@ def handle(self, *args, **options):
+        -if it's greater than the reverse grace, then PINGED QUITE OFTEN
+        -equalise the "previous_ping" and "last_ping" to avoid repeating the PINGED QUITE OFTEN message
+        """
+        try:
```
What's the purpose of the try block if the except part is simply returning?

RrNn replied:
The `try block` is to avoid errors because when a check has not yet pinged, its value of `previous_ping` in the database is `null` and the subtraction of `last_ping` and `previous_ping` throws an exception which I want to catch and simply return, since at this point ,clearly the ping hasn't been pinged many times than it should.

john555 commented on hc/api/models.py:
```diff
@@ -49,6 +49,7 @@ class Meta:
     timeout = models.DurationField(default=DEFAULT_TIMEOUT)
     grace = models.DurationField(default=DEFAULT_GRACE)
     n_pings = models.IntegerField(default=0)
+    previous_ping = models.DateTimeField(null=True,blank=True)
```
What's the difference between `last_ping` and `previous_ping`?

RrNn replied:
`previous_ping` is the time it pinged before the current `last_ping` value, so, when a check pings, I get its `last_ping` value and store it in the `previous_ping` before updating its `last_ping` value, that way I can easily subtract `last_ping` from `previous_ping` to determine whether it has pinged many times than it should

john555 commented on hc/api/views.py:
```diff
@@ -11,17 +11,23 @@
     except Check.DoesNotExist:
         return HttpResponseBadRequest()
+    # get the last_ping to update the previous_ping
+    previous_ping_time = check.last_ping
```
This variable naming is confusing considering you have defined `previous_ping` and `last_ping` separately. Variable names should be intuitive.

RrNn replied:
I'm gonna get a better name for `previous_ping`, one that's more intuitive

RrNn replied:
I've renamed `previous_ping` to `ping_before_last_ping`

john555 commented:
Write tests for the `pinged_often` function.

RrNn committed:
[feature #158175316] Notify user for often pinged checks

Add ping_before_last_ping column to the checks table
Add pinged_often method to sendalerts.py file

[Delivers #158175316]

john555 commented:
@RrNn Resolve conflicts
